generate metadata for requiring transition from beta

deads2k opened the pull request:
This generates the metadata required for requiring the transition from beta (https://github.com/kubernetes/enhancements/tree/master/keps/sig-architecture/1635-prevent-permabeta).  It takes generation tags and creates golang files.

Tags look like this on a type that contains TypeMeta

```
// +k8s:prerelease-lifecycle-gen:introduced=1.1
// +k8s:prerelease-lifecycle-gen:deprecated=1.2
// +k8s:prerelease-lifecycle-gen:removed=1.18
```

If only the introduced version is specified, the other two are defaulted to the policy which reduces toil and prevents cases where people forget to specify.

We can separately add checking to ensure that all beta API folders contain the generation directive.

This starts with just extensions/v1beta1 to vet the methods that are created.  They could later be used by a REST handler and a custom interface to pro grammatically gather the versions.

/kind cleanup
/priority important-soon
@kubernetes/sig-api-machinery-misc 

```release-note
NONE
```

lavalamp commented:
/assign
/assign @liggitt

liggitt commented on staging/src/k8s.io/code-generator/cmd/status-gen/status-generators/status.go:
```diff
@@ -0,0 +1,442 @@
+	}
+
+	if versionedMethodOrDie("Introduced", t) == nil {
+		sw.Do("// Introduced is an autogenerated function, returning the Kubernetes release in which the API struct was introduced and int versions of major and micro for comparison.\n", args)
```
```suggestion
		sw.Do("// Introduced is an autogenerated function, returning the Kubernetes release in which the API struct was introduced and int versions of major and minor for comparison.\n", args)
```

liggitt replied:
is it worth noting the comment format that controls this in a comment in these methods to save someone looking it up?

deads2k replied:
done

liggitt commented on staging/src/k8s.io/api/extensions/v1beta1/doc.go:
```diff
@@ -17,5 +17,6 @@ limitations under the License.
 // +k8s:deepcopy-gen=package
 // +k8s:protobuf-gen=package
 // +k8s:openapi-gen=true
+// +k8s:status-gen=true
```
I think most people will expect this to do something related to a status subresource... not sure what would be clearer... lifecycle or something?

deads2k replied:
> I think most people will expect this to do something related to a status subresource... not sure what would be clearer... lifecycle or something?

`beta-lifecycle-gen`?

liggitt commented on staging/src/k8s.io/code-generator/cmd/status-gen/status-generators/status.go:
```diff
@@ -0,0 +1,442 @@
+// with near certainty when this lifecycle happens as the API is introduced.
+const (
+	tagEnabledName    = "k8s:status-gen"
+	introducedTagName = tagEnabledName + ":introduced"
+	deprecatedTagName = tagEnabledName + ":deprecated"
+	removedTagName    = tagEnabledName + ":removed"
```
do these always refer to kubernetes-apiserver versions? since this is put in the generic code-generator repo, trying to understand how it would interact with the following setups:
1. other API servers building other types outside kube-apiserver and specifying version numbers unrelated to kube
2. API servers serving kube types in a server with a different view of its own version number

deads2k replied:
> do these always refer to kubernetes-apiserver versions? since this is put in the generic code-generator repo, trying to understand how it would interact with the following setups:
> 
> 1. other API servers building other types outside kube-apiserver and specifying version numbers unrelated to kube
> 2. API servers serving kube types in a server with a different view of its own version number

Generator is opt-in, but if someone chose to opt-in, they would use their own version numbers, though major.minor is fixed.  Similar to kube, they would have to provide version.Get results to make decisions.

smarterclayton commented on staging/src/k8s.io/api/extensions/v1beta1/types.go:
```diff
@@ -475,6 +487,9 @@ type DaemonSetCondition struct {
 
 // +genclient
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:status-gen:introduced=1.1
```
Will this show up in godoc?  I really want this in godoc and api doc...

deads2k replied:
> Will this show up in godoc? I really want this in godoc and api doc...

godoc, no because it's in the comment section above like other generator tags

api doc, not yet, but maybe as a future step/verifier.  It *should* be possible to load openapi and diff against the metadata we generated.  This would provide the information to diff against, but I don't want this generator trying to mutate openapi. The metadata to the handler is more valuable at the moment.

deads2k committed:
extension API tag update

deads2k committed:
add prerelease-lifecycle generator for beta API types

deads2k committed:
update makefile for prerelease-lifecycle

deads2k committed:
generated

deads2k commented:
Updated the tags to 

```
// +k8s:prerelease-lifecycle-gen:introduced=1.1
// +k8s:prerelease-lifecycle-gen:deprecated=1.2
// +k8s:prerelease-lifecycle-gen:removed=1.18
```

Update comment in generated code to point to tags in types.go

liggitt commented on staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/prerelease-lifecycle-generators/status.go:
```diff
@@ -0,0 +1,447 @@
+		sw.Do("}\n\n", nil)
+	}
+	if versionedMethodOrDie("Removed", t) == nil {
+		sw.Do("// Removed is an autogenerated function, returning the Kubernetes release in which the API struct was or will be removed and int versions of major and minor for comparison.\n", args)
```
```suggestion
		sw.Do("// Removed is an autogenerated function, returning the release in which the API struct is no longer served, and int versions of major and minor for comparison.\n", args)
```

liggitt replied:
distinguish "no longer served" from "API struct type is removed"? also, if this will generate functions for arbitary API servers' types, the major/minor version might not be a Kubernetes release version, right?

liggitt replied:
is there a reason not to just return the major/minor int versions and drop the string retval?

liggitt replied:
(comments apply to all methods)

deads2k commented:
Jordan confirmed the tag names and generated functions were ok via slack: "gen tags seem ok".

@damemi do a first pass on code please.

damemi reviewed (approved):
/lgtm

deads2k commented:
/retest

liggitt commented:
/approve
